# TODO: Guided haplotype learnning
# 
# Author: xzhou
###############################################################################


source("../rshc/SHC.R")


#' read 1 2 encoded geno data generated by python
#'
readHapData <- function(hapFileName =  "../data/sim_4000seq/80SNP_CEU_sim_4000seq.12encode", popSize, nSnps)
{
	hapData <- read.table(file = hapFileName)
	
	m <- nrow(hapData)
	n <- ncol(hapData)
	
	if(m < popSize || n < 2*nSnps )
	{
		warning("not enough data\n")
		stop()
	}
	
	hapData <- hapData[1:popSize, 1:2*nSnps]
	
	hapData
}

#' evaluate
#' @param targetR the target
#' @param 
#' @param 
#' @param K the weight of rs error and frequency error
hapEvaluate <- function(targetR, targetF, learnedR, learnedF, K = 0.7)
{
	targetRS <- targetR * targetR
	learnedRS <- learnedR * learnedR
	
	rsDiff <- targetRS - learnedRS
	rError <- sum(rsDiff*rsDiff, na.rm = T)/2.0
	
	fDiff <- targetF - learnedF
	fError <- sum(fDiff*fDiff, na.rm = T)/2.0
	
	error <- (1-K)*rError + K*fError
	
	error

}

#' caculate the grad of delta(E)/delta(C) 
grad <- function()
{
	
}

#' learn haplotype
#' 
#' @param targetHaplotype for simulation, we assume we know the phased genotype
#' 			it remembers the genotype
#' @param popSize	The population size of the target haplotype
#' @param nSnps		the number of snps
guidedHapLearning <- function(targetHaplotype, popSize = 20, nSnps = 77)
{
	#calculate the r rs and single allele frequency
	targetR <- calcualteRealR(targetHaplotype)
	targetF <- calculateSingleAlleleFrequence(targetHaplotype)

}






